### PR TITLE
Cancel Points

### PR DESCRIPTION
**Pull request by rennergade**
Adds rust implementation of cancel points for exit/cancel teardown system. Adds additional RPC interface for these calls and datastructures for cancellation.

### CODE REVIEW

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -2277,6 +2277,11 @@ impl Cage {
                 let clonedmutex = mutextable[mutex_handle  as usize].as_ref().unwrap().clone();
                 drop(mutextable);
                 let retval = clonedcv.wait(&*clonedmutex);
+
+                if self.cancelstatus.load(interface::RustAtomicOrdering::Relaxed) {
+                    loop { interface::cancelpoint(self.cageid); } // we check cancellation status here without letting the function return
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:2282]: explain why in a comment

File: src/safeposix/syscalls/net_calls.rs [hunk 2]
```diff
@@ -606,6 +606,10 @@ impl Cage {
                                         //should thus not treat this as a failure in our emulated
                                         //socket; see comment in Socket::new in interface/comm.rs
                                         if sockfdobj.flags & O_NONBLOCK == 0 && i == Errno::EAGAIN {
+                                            if self.cancelstatus.load(interface::RustAtomicOrdering::Relaxed) {
+                                                loop { interface::cancelpoint(self.cageid); } //check cancellation
```
Comment (jesings) [inline src/safeposix/syscalls/net_calls.rs:610]: same as above

**Review by jesings — APPROVED** [review]
Overall looks great, could use a few more comments

**Comment by rennergade** [comment]
@jesings I've updated the comments here, and believe its clearer.

**Comment by jesings** [comment]
Ok I think this is fine.

**Pull request merged**